src/Bicep.Core/Emit/ForSyntaxValidatorVisitor.cs

```diff
@@ -237,20 +283,78 @@ protected override void VisitInternal(SyntaxBase node)
+                        && this.loopLevel == 0
+                        && (this.insideProperties
+                            || this.currentOutputDeclarationSyntax != null
+                            || (this.currentVariableDeclarationSyntax != null && this.variableAccessForInlineCheck == null));
+
+                    if (!isValidResourceCollectionDirectAccessLocation)
+                    {
+                        if (this.variableAccessForInlineCheck != null)
+                        {
+                            IEnumerable<string>? inlineVariableChain = null;
+                            if (this.inlineVariableChainBuilder != null)
+                            {
+                                this.inlineVariableChainBuilder.Add(symbol.Name);
+                                inlineVariableChain = this.inlineVariableChainBuilder.ToImmutable();
+                            }
+                            WriteDirectAccessToCollectionNotSupported(this.variableAccessForInlineCheck, inlineVariableChain);
+                        }
+                        else
+                        {
+                            WriteDirectAccessToCollectionNotSupported(variableOrResourceAccessSyntax);
+                        }
+                    }
+                }
+            } else if ((this.currentVariableDeclarationSyntax == null || this.loopLevel > 0)
+                && variableOrResourceAccessSyntax is VariableAccessSyntax variableAccessSyntax
+                && symbol is VariableSymbol variableSymbol
+                && InlineDependencyVisitor.ShouldInlineVariable(this.semanticModel, variableSymbol.DeclaringVariable, out var outInlineVariableChain))
+            {
+                if (this.inlineVariableVisitLevel == 0)
+                {
+                    this.variableAccessForInlineCheck = variableAccessSyntax;
+                    var chainBuilder = ImmutableArray.CreateBuilder<string>();
+                    chainBuilder.AddRange(symbol.Name);
+                    chainBuilder.AddRange(outInlineVariableChain);
+                    this.inlineVariableChainBuilder = chainBuilder;
+                }
+
+                this.inlineVariableVisitLevel++;
+                VisitVariableDeclarationSyntax(variableSymbol.DeclaringVariable);
+                this.inlineVariableVisitLevel--;
+
+                if (this.inlineVariableVisitLevel == 0)
+                {
+                    this.variableAccessForInlineCheck = null;
+                    this.inlineVariableChainBuilder = null;
                 }
             }
         }
 
-        private static ObjectPropertySyntax? TryGetDependsOnProperty(ObjectSyntax? body) => body?.TryGetPropertyByName("dependsOn");
+        private void WriteDirectAccessToCollectionNotSupported(IPositionable positionable, IEnumerable<string>? accessChain = null) =>
+            this.diagnosticWriter.Write(DiagnosticBuilder.ForPosition(positionable).DirectAccessToCollectionNotSupported(accessChain));
```

Comment:
One trick we can do here is to split the method into two overloads:

```C#
private void WriteDirectAccessToCollectionNotSupported(IPositionable positionable) =>
    WriteDirectAccessToCollectionNodeSupported(positionable, Enumerable.Empty<string>());

private void WriteDirectAccessToCollectionNotSupported(IPositionable positionable, IEnumerable<string> accessChain) =>
    ...
```

This would save the null check in `DirectAccessToCollectionNotSupported`.